### PR TITLE
migrate python structure tests to structure test framework

### PR DESCRIPTION
**Pull request by nkubala**
This will migrate all tests from the Python runtime image testsuite that fall under the "structure test" category over to running on top of the GCP structure test framework.

Because of this project's heavy dependencies on `make`, I'm leaving the legacy tests in the codebase. We should make an effort to unify the build process for this image as soon as possible; since cloudbuild was recently integrated into the Makefile, it seems this is the direction things are headed. We'll just need to make sure we're not explicitly building the image and then running tests through `make`, as this would bypass the structure test cloudbuild step which actually runs the tests.

The remaining tests (`benchmark` and `google-cloud-python`) will be covered under the planned integration testing framework.

dependent on https://github.com/GoogleCloudPlatform/runtimes-common/pull/61

@dlorenc  @duggelz @jonparrott

### CODE REVIEW

**Comment by jonparrott** [comment]
@nkubala this looks neat. I have a few questions:

1. What's the biggest gain of using this over just making cloud build run `docker build` against our current test dockerfiles?
2. How would user run this locally (e.g., without cloud builder)?

**Comment by nkubala** [comment]
@jonparrott good questions.

> What's the biggest gain of using this over just making cloud build run docker build against our current test dockerfiles?

We're making a big effort to unify the test/build pipeline across all the silver language runtimes. This framework is one of a few different build steps that we're recommending silver language teams use in order to standardize the way we do CI across the runtimes. As far as actual functional differences go between this and doing a `docker build` on your test files, this will provide you with significantly more verbose logging, as well as the ability to continue running tests after one fails (whereas dockerfiles would just exit immediately). It also provides an easy way to run more complex tests, such as checking the error code of different commands, or examining files on the file system. Not to say this isn't doable through a bash script; this just makes it painless for the test writers.

> How would user run this locally (e.g., without cloud builder)?

We have a script that will run these in a very similar way to the cloudbuild run, but can be run on any host machine that has a working docker installation.
```
curl https://raw.githubusercontent.com/GoogleCloudPlatform/runtimes-common/master/structure_tests/ext_run.sh > ext_run.sh
bash ext_run.sh -i gcr.io/gcp-runtimes/check_if_tag_exists -c check_if_image_tag_exists/test_config.json
```

Check out the full documentation for running these [here](https://github.com/GoogleCloudPlatform/runtimes-common/tree/master/structure_tests).

**Comment by jonparrott** [comment]
@nkubala thanks! Some suggestions based on your comments.

1. Remove the test dockerfiles altogether.
2. Update the makefiles to have a `local` and `cloud` tasks. Local should use the script you mentioned, cloud should kickoff the builder.

**Comment by nkubala** [comment]
> Remove the test dockerfiles altogether.

Didn't want to do this until I got the ok, glad you're on board though :)

> Update the makefiles to have a local and cloud tasks. Local should use the script you mentioned, cloud should kickoff the builder.

This is doable, but there will be a few differences. The most important one is that kicking off the cloudbuilder will actually build the image, and then run the tests on that new image candidate; kicking off the local tests will just run the tests on a specified image. So this means any time you run `make local`, you should probably have already done at least a `make build`; not sure if it makes sense to make `build` a dependency of `local` in the makefile though (since we probably don't want to rebuild **every** time we run tests locally). Just something to be aware of.

**Comment by jonparrott** [comment]
Cool, let's do this instead: only the top-level makefile has rules to do cloud build. The test makefiles just have the rules to do local stuff.

**Comment by nkubala** [comment]
Sounds good. So `make test` from top-level will essentially be local testing then?

**Comment by jonparrott** [comment]
@nkubala I think so, maybe it makes sense to do:

* `local-image`.
* `local-tests: local-image`
* `build` - cloud build, does build and tests and everything.

@duggelz should definitely have some input here on how granular he wants the cloud build step to be.

File: tests/virtualenv/virtualenv.yaml [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    command: ["python", "-c", "\"import sys; import flask; sys.exit(0 if flask.__file__.startswith('/env') else 1)\""]
+
+#####################################################
+#    Python 3.5 Tests
+#####################################################
+
+  - name: "virtual env teardown"
+    command: ["rm", "-rf", "/env"]
+
+  - name: "virtualenv python installation"
```
Comment (duggelz) [inline tests/virtualenv/virtualenv.yaml]: Add something like lines 49-51 here:

    - name: "python installation"
      command: ["which", "python3.5"]
      expectedOutput: ["/opt/python3.5/bin/python3.5\n"]

And change the virtualenv call to

      setup: [["virtualenv", "-p", "python3.5", "/env"]]

File: tests/python3-libraries/python3-libraries.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  - name: "env var setup"
+    envVars:
+      - key: "VIRTUAL_ENV"
+        value: "/env"
+      - key: "PATH"
+        value: "/env/bin:$PATH"
+    command: ["echo", "'setup complete'"]
+
+  - name: "requirements"
+    setup: [["virtualenv", "-p", "/opt/python3.5/bin/python3.5", "/env"]]
```
Comment (duggelz) [inline tests/python3-libraries/python3-libraries.yaml]: I added python3.5 to the $PATH, so change "/opt/python3.5/bin/python3.5" to "python3.5" here.

**Comment by duggelz** [comment]
Actually, we only need one of "External users can run the cloud build and test pipeline, but it puts the image somewhere else" and "make local-test", not both, so I retract # 2.

**Comment by duggelz** [comment]
Ok, rereading the Container Builder docs, the image doesn't go to GCR until the tests pass, so I retract # 1 as well.

File: tests/virtualenv/virtualenv.yaml (hunk 1 above)
Comment (nkubala) [inline tests/virtualenv/virtualenv.yaml]: done.

File: tests/python3-libraries/python3-libraries.yaml (hunk 2 above)
Comment (nkubala) [inline tests/python3-libraries/python3-libraries.yaml]: done.

**Comment by nkubala** [comment]
ok, I updated the targets in the Makefile based on the suggestions. the only thing left out of them currently is running benchmark/gcloud-system tests from the cloudbuild; the reason for this being the structure tests are integrated as an intermediate step in the cloudbuild, but since these are integration tests we don't have a good way to run them through the cloudbuild currently. this will come as part of the integration testing framework I have planned in the next month or so.

**Comment by nkubala** [comment]
@duggelz @jonparrott I've updated a few things related to the build process; please take one more look and make sure things look alright.

* `build`: cloudbuild (with implicit structure tests) and integration tests
* `build-local`: local image build, local structure tests, integration tests
* `integration-tests`: benchmarks, google-cloud-python and google-cloud-system-tests (these can't yet be run in cloudbuild)
* `structure-tests`: local structure tests using external run script.

I also parameterized the Dockerfiles to take in the tag for the image that just got built; this lets us run the integration tests on the newly pushed image (since assuming the structure tests in cloudbuild pass, the image will have just been pushed to gcr.io).

**Comment by nkubala** [comment]
once this is merged, I'll update the Jenkins job to just call the `jenkins_build.sh` script (which is a very slightly modified version of what's being run to build now)

**Comment by duggelz** [comment]
Looks good to me.

**Pull request merged**